### PR TITLE
[openflow] Openflow group tables support

### PR DESCRIPTION
**Pull request by pierrecregut**
The Openflow probe did not display group tables because there was no obvious way to monitor them. Those two commits add support for groups. The hard part is in the probe where we use a mostly hidden functionality of OpenFlow 1.4. As one can expect group support will only work for OF1.5 enabled switch (we cannot safely enable it for 1.4 due to a bug in openvswitch corrected last week).

### CODE REVIEW

File: topology/probes/ovsdb/ovsdb.go [hunk 1]
```diff
@@ -129,6 +127,7 @@ func (o *OvsdbProbe) OnOvsBridgeAdd(monitor *ovsdb.OvsMonitor, uuid string, row
 			}
 		}
 	}
+	o.Graph.Unlock()
```
Comment (lebauce) [inline topology/probes/ovsdb/ovsdb.go]: The graph is still used in `OnOvsBridgeAdd` so should we stay under lock. Maybe we should just Unlock and RLock

File: topology/probes/ovsdb/ovs_of.go [hunk 2]
```diff
@@ -556,10 +826,11 @@ func (o *OvsOfProbe) OnOvsBridgeAdd(bridgeNode *graph.Node) {
 	bridgeName, _ := bridgeNode.GetFieldString("Name")
 	bridgeProbe, err := o.NewBridgeProbe(hostName, bridgeName, uuid, bridgeNode)
 	if err == nil {
-		logging.GetLogger().Infof("Probe added for %s on %s (%s)", bridgeName, hostName, uuid)
+		logging.GetLogger().Debugf("Probe added for %s on %s (%s)", bridgeName, hostName, uuid)
 		o.BridgeProbes[uuid] = bridgeProbe
 	} else {
 		logging.GetLogger().Errorf("Cannot add probe for bridge %s on %s (%s)", bridgeName, hostName, uuid)
```
Comment (lebauce) [inline topology/probes/ovsdb/ovs_of.go]: logging.GetLogger().Errorf("Cannot add probe for bridge %s on %s (%s) : %s", bridgeName, hostName, uuid, err)

File: topology/probes/ovsdb/ovs_of.go [hunk 3]
```diff
@@ -521,9 +781,19 @@ func (o *OvsOfProbe) NewBridgeProbe(host string, bridge string, uuid string, bri
...
 
-	return probe, probe.monitor(ctx)
+	err := probe.monitor(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	err = probe.monitorGroup(ctx)
+	if err != nil {
+		logging.GetLogger().Errorf("Cannot add group probe on %s - %s", bridge, err.Error())
```
Comment (lebauce) [inline topology/probes/ovsdb/ovs_of.go]: nit: s/err.Error()/err/

File: topology/probes/ovsdb/ovsdb.go (hunk 1 above)
Comment (pierrecregut) [reply topology/probes/ovsdb/ovsdb.go]: I have put locks on "atomic" graph operations like adding a node. OnOvsBridgeAdd does much more than that and locks are not recursive in go. So I think releasing the lock is sensible but using an explicit unlock is more error prone than defer. Another solution would be to push the unlock defer rlock to the callback but relocking for an immediate unlock is weird.

**Comment by pierrecregut** [comment]
run skydive-ppc64le-tests

**Comment by pierrecregut** [comment]
run skydive-functional-tests-backend-elasticsearch insanelock

**Comment by pierrecregut** [comment]
run skydive-functional-tests-backend-elasticsearch  insanelock

File: topology/probes/ovsdb/ovsdb.go (hunk 1 above)
Comment (pierrecregut) [reply topology/probes/ovsdb/ovsdb.go]: Functional tests detected a deadlock due to recursive use of graph lock. Invariants maintained here: graph is only locked on elementary actions related to it (no system stuff done while the lock is held). openflow probe lock must be locked *before* graph lock

Comment (pierrecregut) [reply topology/probes/ovsdb/ovsdb.go]: Race removed when we add  a group just after the creation of the bridge. Takes also into account the fact that the bridge does not recognized OF1.5 when it is created, only after a set bridge protocols=...

**Comment by pierrecregut** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by pierrecregut** [comment]
run skydive-k8s-tests

File: topology/probes/ovsdb/ovs_of.go [hunk 4]
```diff
@@ -329,7 +386,7 @@ func launchContinuousOnSwitch(ctx context.Context, cmd []string) (<-chan string,
 
 				return nil
 			}
-			if err := common.Retry(retry, 100, 10*time.Millisecond); err != nil {
+			if err := common.Retry(retry, 100, 500*time.Millisecond); err != nil {
```
Comment (safchain) [inline topology/probes/ovsdb/ovs_of.go]: 50 sec seems to be too long ? Is there anything special requiring to wait so long time ?

File: topology/probes/ovsdb/ovs_of.go [hunk 5]
```diff
@@ -506,6 +686,112 @@ func (probe *BridgeOfProbe) monitor(ctx context.Context) error {
 	return nil
 }
 
+const delayForFile = 30
+
+// Check if a file is created in time
+func waitForFile(path string) error {
```
Comment (safchain) [inline topology/probes/ovsdb/ovs_of.go]: should be replace by comm.Retry

File: topology/probes/ovsdb/ovs_of.go [hunk 6]
```diff
@@ -506,6 +686,112 @@ func (probe *BridgeOfProbe) monitor(ctx context.Context) error {
...
+	}
+	return err
+}
+
+// monitorGroup monitors the openflow groups of a bridge by
+// launching a goroutine. It uses OpenFlow 1.4 ForwardRequest
+// command that is not directly available in ovs-ofctl
+//
+// Note: Openflow15 is really needed. Receiving an insert_bucket on an OF14
+// monitor crashes the switch (yes, the switch itself) on OVS 2.9
```
Comment (safchain) [inline topology/probes/ovsdb/ovs_of.go]: I like this comment :) it is scary but funny

File: topology/probes/ovsdb/ovs_of.go [hunk 7]
```diff
@@ -548,23 +846,27 @@ func (o *OvsOfProbe) OnOvsBridgeAdd(bridgeNode *graph.Node) {
...
 	hostName := o.Host
 	bridgeName, _ := bridgeNode.GetFieldString("Name")
 	bridgeProbe, err := o.NewBridgeProbe(hostName, bridgeName, uuid, bridgeNode)
 	if err == nil {
-		logging.GetLogger().Infof("Probe added for %s on %s (%s)", bridgeName, hostName, uuid)
+		logging.GetLogger().Debugf("Probe added for %s on %s (%s)", bridgeName, hostName, uuid)
 		o.BridgeProbes[uuid] = bridgeProbe
 	} else {
 		logging.GetLogger().Errorf("Cannot add probe for bridge %s on %s (%s)", bridgeName, hostName, uuid)
+		logging.GetLogger().Error(err.Error())
```
Comment (safchain) [inline topology/probes/ovsdb/ovs_of.go]: nit: logging.GetLogger().Error(err)

File: topology/probes/ovsdb/ovs_of.go (hunk 4 above)
Comment (pierrecregut) [reply topology/probes/ovsdb/ovs_of.go]: No but 1s may be too short on a slow machine if multiple bridges are created. This is the propagation delay between the admin state of the database and the real bridge in the switch daemon. Let us choose 5s

File: topology/probes/ovsdb/ovs_of.go (hunk 7 above)
Comment (pierrecregut) [reply topology/probes/ovsdb/ovs_of.go]: and could be joined with previous line.

**Pull request merged**